x/pool-incentives/keeper/grpc_query.go

```diff
@@ -143,54 +143,45 @@ func (q Querier) IncentivizedPools(ctx context.Context, _ *types.QueryIncentiviz
 	// While there are exceptions, typically the number of incentivizedPools
 	// equals to the number of incentivized gauges / number of lockable durations.
 	incentivizedPools := make([]types.IncentivizedPool, 0, len(distrInfo.Records)/len(lockableDurations))
-	incentivizedPoolIDs := make(map[uint64]time.Duration)
 
 	// Loop over the distribution records and fill in the incentivized pools struct.
 	for _, record := range distrInfo.Records {
-		for _, lockableDuration := range lockableDurations {
-			poolId, err := q.Keeper.GetPoolIdFromGaugeId(sdkCtx, record.GaugeId, lockableDuration)
-			if err == nil {
-				incentivizedPool := types.IncentivizedPool{
-					PoolId:           poolId,
-					LockableDuration: lockableDuration,
-					GaugeId:          record.GaugeId,
-				}
-
-				incentivizedPools = append(incentivizedPools, incentivizedPool)
-				incentivizedPoolIDs[poolId] = lockableDuration
-			}
-		}
-	}
-
-	// Only run the following if the above loop determined there were incentivized pools.
-	if len(incentivizedPoolIDs) > 0 {
-		// Retrieve the migration records between balancer pools and concentrated liquidity pools.
-		// This comes from the superfluid keeper, since superfluid is the only pool incentives connected
-		// module that has access to the gamm modules store.
-		migrationRecords, err := q.gammKeeper.GetAllMigrationInfo(sdkCtx)
+		gauge, err := q.incentivesKeeper.GetGaugeByID(sdkCtx, record.GaugeId)
 		if err != nil {
 			return nil, status.Error(codes.Internal, err.Error())
 		}
 
-		// Iterate over all migration records.
-		for _, record := range migrationRecords.BalancerToConcentratedPoolLinks {
-			// If the cl pool is not in the list of incentivized pools, skip it.
-			lockableDuration, incentivized := incentivizedPoolIDs[record.ClPoolId]
-			if !incentivized {
-				continue
+		if gauge.DistributeTo.LockQueryType == lockuptypes.ByGroup {
+			group, err := q.Keeper.incentivesKeeper.GetGroupByGaugeID(sdkCtx, record.GaugeId)
+			if err != nil {
+				return nil, status.Error(codes.Internal, err.Error())
 			}
-
-			// Add the indirectly incentivized balancer pools to the list of incentivized pools.
-			gaugeId, err := q.Keeper.GetPoolGaugeId(sdkCtx, record.BalancerPoolId, lockableDuration)
-			if err == nil {
+			poolIds, durations, err := q.Keeper.incentivesKeeper.GetPoolIdsAndDurationsFromGroup(sdkCtx, group)
+			if err != nil {
+				return nil, status.Error(codes.Internal, err.Error())
+			}
+			for i, poolId := range poolIds {
 				incentivizedPool := types.IncentivizedPool{
-					PoolId:           record.BalancerPoolId,
-					LockableDuration: lockableDuration,
-					GaugeId:          gaugeId,
+					PoolId:           poolId,
+					LockableDuration: durations[i],
+					GaugeId:          record.GaugeId,
 				}
 
 				incentivizedPools = append(incentivizedPools, incentivizedPool)
 			}
+		} else {
+			for _, lockableDuration := range lockableDurations {
```

Comment:
I think this is a mistake from the past but we also need a branch for handling NoLock gauges. NoLock gauges do not need to iterate over all lockable durations. I suspect that for them, we return the same pool 3 times.

Do we have a test reproducing this? Do you mind addressing this issue in this PR?

Reply:
Added separate logic gates here https://github.com/osmosis-labs/osmosis/pull/6601/commits/8dc2d76814df091594afe2e27d003b212c665ce0

The behavior you described never happened for the following reason:

When we create a noLock gauge, we create a link between epoch time and the gauge itself (see https://github.com/osmosis-labs/osmosis/blob/7fd52e848d1de0c75bcba221d8bee9599da33669/x/pool-incentives/keeper/keeper.go#L129-L132)

Therefore, when we looped over all the durations, we would pull the record for the noLock gauge when we hit the epoch duration, and for the other durations we would fail silently.

Reply:
Sweet, thanks for digging into this!